<launch>

    <!-- SACSegmemtation to remove ground plane -->
    <node name="floor_removal" pkg="agilex_scout_mini" type="floor_removal" output="screen"/>

    <!-- Transforms data from dr_spaam to People -->
    <!--node name="pose_array_to_people" pkg="agilex_scout_mini" type="pose_array_to_people" output="screen"/-->

    <!-- Launch the nodelet manager for pcl -->
    <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen" />

    <!-- Passthrough to filter out points outside the CropBox -->
    <node pkg="nodelet" type="nodelet" name="passthrough" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input" to="/velodyne_points/floor_removed/" />
        <remap from="~output" to="/velodyne_points/passthrough" />
        <rosparam>
            filter_field_name: x
            filter_limit_min: -20.00
            filter_limit_max: 20.0
            filter_limit_negative: False

            filter_field_name: y
            filter_limit_min: -20.0
            filter_limit_max: 20.0
            filter_limit_negative: False
        </rosparam>
    </node>

    <!-- Get rid of pesky noise using radius outlier filter -->
    <node pkg="nodelet" type="nodelet" name="radius_outlier_removal" args="load pcl/RadiusOutlierRemoval pcl_manager" output="screen">
        <remap from="~input" to="/velodyne_points/passthrough" />
        <remap from="~output" to="/velodyne_points/outlier_removed" />
        <rosparam>
            <!-- Neighbor area to check points -->
            <!-- Keep points with neighbor points larger than this number -->
            radius_search: 0.20
            min_neighbors: 1
        </rosparam>
    </node>

    <!-- VoxelGrid downsampling to produce a less dense point cloud -->
    <node pkg="nodelet" type="nodelet" name="voxel_grid" args="load pcl/VoxelGrid pcl_manager" output="screen">
        <remap from="~input" to="/velodyne_points/outlier_removed" />
        <remap from="~output" to="/velodyne_points/filtered" />
        <rosparam>
            filter_field_name: z
            filter_limit_min: -0.25
            filter_limit_max: 0.3
            filter_limit_negative: False
            leaf_size: 0.05
        </rosparam>
    </node>
    
    <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
        <remap from="cloud_in" to="/velodyne_points/filtered"/>    
        <remap from="scan" to="front_scan"/>
        <rosparam>
            transform_tolerance: 0.01
            min_height: 0.0
            max_height: 1.0

            angle_min: -3.1416
            angle_max: 3.1416
            angle_increment: 0.0087
            scan_time: 0.05
            range_min: 0.1
            use_inf: false
            concurrency_level: 1
        </rosparam>
    </node>
</launch> 
